### PR TITLE
[CFP-69] FormBuilder migration- Supporting evidence

### PR DESCRIPTION
**Pull request by NenehB**
#### What
Migrate the relevant form to use the [GOVUKDesignSystemFormBuilder](https://govuk-form-builder.netlify.app/) gem.

#### Ticket

[CFP-69](https://dsdmoj.atlassian.net/browse/CFP-69)

#### Why

- Continue the migration away from the deprecated design system to the current GOVUK Frontend
- Resolve related accessibility issues identified in the [2020 DAC Accessibility report](https://drive.google.com/file/d/16RrVZoZFb8VMSlwUj_kfvwhm-mAVCKXs/view?usp=sharing)
- Reduce/remove redundant form builders throughout the codebase

--------

#### TODO

 - [x] Migrate FormBuilder elements
 - [x] Fix failing tests

### CODE REVIEW

File: app/controllers/external_users/claims_controller.rb [hunk 1]
```diff
@@ -321,14 +321,14 @@ def claim_params
         :claim_id,
         :first_name,
         :last_name,
-        date_attributes_for(:date_of_birth),
+        :date_of_birth,
```
Comment (naseberry) [inline app/controllers/external_users/claims_controller.rb]: This is related to the `defendants` form_step, is this required for `supporting_evidence`?

File: app/controllers/external_users/claims_controller.rb [hunk 2]
```diff
@@ -321,14 +321,14 @@ def claim_params
...
         :order_for_judicial_apportionment,
         :_destroy,
         representation_orders_attributes: [
           :id,
           :document,
           :maat_reference,
-          date_attributes_for(:representation_order_date),
+          :representation_order_date,
```
Comment (naseberry) [inline app/controllers/external_users/claims_controller.rb]: This is related to the `defendants` form_step, is this required for `supporting_evidence`?

File: app/models/defendant.rb [hunk 3]
```diff
@@ -23,10 +23,6 @@ class Defendant < ApplicationRecord
   validates_with DefendantValidator
   validates_with DefendantSubModelValidator
 
-  acts_as_gov_uk_date :date_of_birth,
-                      validate_if: :validate_date?,
-                      error_clash_behaviour: :override_with_gov_uk_date_field_error
-
```
Comment (naseberry) [inline app/models/defendant.rb]: This is related to the `defendants` form_step, is this required for `supporting_evidence`?

File: app/models/representation_order.rb [hunk 4]
```diff
@@ -23,10 +23,6 @@ class RepresentationOrder < ApplicationRecord
     self.maat_reference = nil if case_type&.requires_maat_reference?.eql?(false)
   end
 
-  acts_as_gov_uk_date :representation_order_date,
-                      validate_if: :perform_validation?,
-                      error_clash_behaviour: :override_with_gov_uk_date_field_error
-
```
Comment (naseberry) [inline app/models/representation_order.rb]: This is related to the `defendants` form_step, is this required for `supporting_evidence`?

File: app/validators/defendant_validator.rb [hunk 5]
```diff
@@ -19,9 +19,9 @@ def validate_claim
   end
 
   def validate_date_of_birth
-    validate_presence(:date_of_birth, 'blank')
-    validate_on_or_before(10.years.ago, :date_of_birth, 'check')
-    validate_on_or_after(120.years.ago, :date_of_birth, 'check')
+    validate_presence(:date_of_birth, :blank)
```
Comment (naseberry) [inline app/validators/defendant_validator.rb]: This modified file relates to the `defendants` form_step, is this required for `supporting_evidence`?

File: app/validators/representation_order_validator.rb [hunk 6]
```diff
@@ -14,8 +14,8 @@ def self.fields
   # must not be earlier than the first rep order date
   # must not be earlier than the earliest permitted date
   def validate_representation_order_date
-    validate_presence(:representation_order_date, 'blank')
-    validate_on_or_before(Date.today, :representation_order_date, 'in_future')
+    validate_presence(:representation_order_date, :blank)
```
Comment (naseberry) [inline app/validators/representation_order_validator.rb]: This modified file relates to the `defendants` form_step, is this required for `supporting_evidence`?

File: app/views/external_users/claims/_error_summary.html.haml [hunk 7]
```diff
@@ -1,4 +1,4 @@
-- if %w[case_details transfer_fee_details].include?(@claim.form_step.to_s)
+- if %w[case_details defendants transfer_fee_details supporting_evidence].include?(@claim.form_step.to_s)
```
Comment (naseberry) [inline app/views/external_users/claims/_error_summary.html.haml]: Do we need to define `defendants` for `supporting_evidence` changes?

File: app/views/external_users/claims/supporting_documents/_disk_evidence.html.haml [hunk 8]
```diff
@@ -1,15 +1,4 @@
...
-        = b.radio_button('aria-labelledby': "radio-control-legend-evidence #{b.text.to_css_class}")
-        = b.label(id: b.text.to_css_class) { b.text }
-      - if (b.text == 'Yes')
-        #evidence-info.panel.panel-border-narrow.js-hidden
-          = render partial: 'shared/disk_evidence_info', locals: { f: f }
-
-  = validation_error_message(@error_presenter, :disk_evidence)
+= f.govuk_radio_buttons_fieldset :disk_evidence, legend: { size: 'm', text: t('disk_evidence.question') } do
+  = f.govuk_radio_button :disk_evidence, 'true', label: { text: t('global_yes') }, link_errors: true do
+    = render partial: 'shared/disk_evidence_info', locals: { f: f }
```
Comment (naseberry) [inline app/views/external_users/claims/supporting_documents/_disk_evidence.html.haml]: `locals` is no longer required and can be removed.

File: app/views/external_users/claims/supporting_documents/_new.html.haml [hunk 9]
```diff
@@ -12,9 +12,11 @@
   %fieldset
     %legend
       = t('.legend')
-    .dropzone
-      = f.label :documents, 'Upload file', class: 'form-group'
-      = f.file_field :documents, name: 'documents', multiple: true, class: 'form-control'
+    
+= f.govuk_file_field :documents,
+  form_group: { classes: 'dropzone' },
+  label: { text: 'Upload file' },
```
Comment (naseberry) [inline app/views/external_users/claims/supporting_documents/_new.html.haml]: Minor: add text to relevant locale file

File: app/views/external_users/claims/supporting_evidence_checklist/_expenses_invoices_requirement.html.haml [hunk 10]
```diff
@@ -0,0 +1,8 @@
+%p{ class: 'govuk-body govuk-!-font-weight-bold' }
```
Comment (naseberry) [inline app/views/external_users/claims/supporting_evidence_checklist/_expenses_invoices_requirement.html.haml]: Minor: add content to relevant locale file

File: app/views/external_users/claims/supporting_evidence_checklist/_fields.html.haml [hunk 11]
```diff
@@ -1,7 +1,8 @@
-%fieldset.evidence-checklist{ 'data-mute-indictment': @claim.allows_fixed_fees? ? 'true' : 'false' }
-  %legend.form-label-bold
-    = t('.supporting_evidence_checklist')
-    %span.form-hint
-      = t('.supporting_evidence_checklist_hint')
+= f.govuk_check_boxes_fieldset :evidence_checklist_ids, form_group: { 'data-mute-indictment': @claim.allows_fixed_fees? ? 'true' : 'false' } , legend: { text: t('.supporting_evidence_checklist'), size: 's' } do
```
Comment (naseberry) [inline app/views/external_users/claims/supporting_evidence_checklist/_fields.html.haml]: Input is missing the `hint`

File: app/views/external_users/claims/supporting_evidence_checklist/_fields.html.haml [hunk 12]
```diff
@@ -1,7 +1,8 @@
...
+  - f.object.eligible_document_types.each do |eligible_document_type|
+    = f.govuk_check_box :evidence_checklist_ids,
+      eligible_document_type.id,
+      label: { text: eligible_document_type.name } do
 
-  = render partial: 'external_users/claims/supporting_evidence_checklist/evidence_checklist', locals: { f: f, collection: f.object.eligible_document_types }
+      - if eligible_document_type.name.eql?('Expenses invoices')
+        = render partial: 'external_users/claims/supporting_evidence_checklist/expenses_invoices_requirement', locals: { f: f }
```
Comment (naseberry) [inline app/views/external_users/claims/supporting_evidence_checklist/_fields.html.haml]: `locals` is no longer required and can be removed.

File: app/webpack/javascripts/modules/external_users/claims/NewClaim.js [hunk 13]
```diff
@@ -15,7 +15,7 @@ moj.Modules.NewClaim = {
     // Tests in /spec/javascripts/supporting-evidence_spec.js
     //
     $('button[name="commit_submit_claim"]').on('click', function (e) {
-      if ($('#claim_evidence_checklist_ids_4').exists() && !$('#claim_evidence_checklist_ids_4').prop('checked') && !$('[data-mute-indictment]').data('mute-indictment')) {
+      if ($('#claim-evidence-checklist-ids-4-field').exists() && !$('#claim-evidence-checklist-ids-4-field').prop('checked')&& !$('[data-mute-indictment]').data('mute-indictment')) {
```
Comment (naseberry) [inline app/webpack/javascripts/modules/external_users/claims/NewClaim.js]: Linting: re-apply the space before `&&`

File: app/views/external_users/claims/supporting_documents/_disk_evidence.html.haml (hunk 8 above)
Comment (NenehB) [reply app/views/external_users/claims/supporting_documents/_disk_evidence.html.haml]: Done

File: app/views/external_users/claims/supporting_documents/_new.html.haml (hunk 9 above)
Comment (NenehB) [reply app/views/external_users/claims/supporting_documents/_new.html.haml]: Done

File: app/views/external_users/claims/supporting_evidence_checklist/_expenses_invoices_requirement.html.haml (hunk 10 above)
Comment (NenehB) [reply app/views/external_users/claims/supporting_evidence_checklist/_expenses_invoices_requirement.html.haml]: Done

File: app/views/external_users/claims/supporting_evidence_checklist/_fields.html.haml (hunk 11 above)
Comment (NenehB) [reply app/views/external_users/claims/supporting_evidence_checklist/_fields.html.haml]: Done

File: app/views/external_users/claims/supporting_evidence_checklist/_fields.html.haml (hunk 12 above)
Comment (NenehB) [reply app/views/external_users/claims/supporting_evidence_checklist/_fields.html.haml]: Done

File: app/webpack/javascripts/modules/external_users/claims/NewClaim.js (hunk 13 above)
Comment (NenehB) [reply app/webpack/javascripts/modules/external_users/claims/NewClaim.js]: Done

File: app/views/external_users/claims/supporting_documents/_new.html.haml [hunk 14]
```diff
@@ -12,9 +12,11 @@
   %fieldset
     %legend
       = t('.legend')
```
Comment (naseberry) [inline app/views/external_users/claims/supporting_documents/_new.html.haml]: Orphaned fieldset, consider its usage or change to a heading to reflect the section

File: app/views/external_users/claims/supporting_evidence_checklist/_fields.html.haml [hunk 15]
```diff
@@ -1,7 +1,13 @@
-%fieldset.evidence-checklist{ 'data-mute-indictment': @claim.allows_fixed_fees? ? 'true' : 'false' }
-  %legend.form-label-bold
-    = t('.supporting_evidence_checklist')
-    %span.form-hint
-      = t('.supporting_evidence_checklist_hint')
+= f.govuk_check_boxes_fieldset :evidence_checklist_ids,
+  classes: 'cc-multiple-choice',
```
Comment (naseberry) [inline app/views/external_users/claims/supporting_evidence_checklist/_fields.html.haml]: This class wraps the group, not the individual check boxes, therefore i don't think this is doing what you expected.

File: config/locales/en.yml [hunk 16]
```diff
@@ -1501,6 +1502,15 @@ en:
...
           supporting_evidence_checklist_hint: Select all that apply
+          expenses_invoices_requirement_html:
+            <p class = 'govuk-body govuk-!-font-weight-bold'>
+            Please make sure invoices include:</p>
+            <ul class = 'govuk-list govuk-list--bullet'>
+              <li>date</li>
+              <li>case number or case URN</li>
+              <li> defendant name</li>
+              <li> hourly rates and details of the work done</li>
+            </ul>
```
Comment (naseberry) [inline config/locales/en.yml]: White space needs cleanup

File: features/page_objects/sections/evidence_checklist_section.rb [hunk 17]
```diff
@@ -1,5 +1,5 @@
 class EvidenceChecklistSection < SitePrism::Section
-  sections :items, ".multiple-choice" do
+  sections :items, ".cc-multiple-choice .govuk-checkboxes__item" do
```
Comment (naseberry) [inline features/page_objects/sections/evidence_checklist_section.rb:2]: This class wraps the group, not the individual check boxes, therefore I'm not sure why this is working, let's have a review together.

**Review by naseberry — APPROVED** [review]
LGTM 👍

**Review by jrmhaig — APPROVED** [review]
Looks good. Just the conflict with `app/views/external_users/claims/_error_summary.html.haml` that needs to be resolved.

**Pull request merged**